File: valid-parentheses-20/Solution.java

/**
 * LeetCode problem # 20, 'Valid Parentheses'
 * https://leetcode.com/problems/valid-parentheses
 */

import java.util.*;

class Solution {
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<>();
        Map<Character, Character> parentheses = new HashMap<>();
        parentheses.put(']', '[');
        parentheses.put('}', '{');
        parentheses.put(')', '(');

        for (int i = 0; i < s.length(); ++i) {
            char curr = s.charAt(i);
            if (parentheses.containsKey(curr)) {
                if (st.isEmpty()) {
                    return false;
                }
                char last = (char) st.pop();
                char parenthesis = parentheses.get(curr);
                if (last != parenthesis) {
                    return false;
                }
            } else {
                st.push(curr);
            }
        }
        return st.isEmpty();
    }
}

File: valid-parentheses-20/solution.ts

/**
 * LeetCode problem # 20, 'Valid Parentheses'
 * https://leetcode.com/problems/valid-parentheses
 */
function isValid(s: string): boolean {
    const stack: string[] = [];
    const parens = new Map([
        [']', '['],
        ['}', '{'],
        [')', '('],
    ]);

    for (const char of s) { 
        const p = parens.get(char);
        if (p) {
            if (!stack.length) {
                return false;
            }
            const last = stack.pop();
            if (last !== p) {
                return false;
            }
        } else {
            stack.push(char);
        }
    }
    return stack.length === 0;
}